xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,355 @@
+# recall that in xDSL an operation can contain regions which in turn can contain
+# CFGs of basic blocks. An operation passes control flow to a region and on exit
+# from the region the control flow returns to the operation. Then the operation
+# can transfer control flow to another region, etc.
+#
+# Desymrefication can be applied to regions and operations. First, consider a
+# simple case when all operations in the region do not have any nested regions.
+# This means that the region is simply a CFG. Otherwise, there are some
+# operations that contain regions. However, we can apply desymrefication to
+# these operations them first, reducing the problem to the the first case. In
+# pseudocode, that would look like:
+#
+# ```
+# desymref_region(region):
+#   for block in region.blocks:
+#     for op in block.ops:
+#       desymref_op(op)
+#     # at this point no op contains a symref operation
+#     # in the nested region.
+# ```
+#
+# Next, we describe how desymrefication actually works, on a high level.
+# Every region either declares a symbol or uses it from the parent region.
+# Let's consider each case separately.
+#
+# Case 1: Symbol is declared in this region. This is an easy case. We already
+# know that symbol is only used in this region and no other nested region uses
+# it. Therefore running any SSA-construction algorithm on the CFG is enough.
+#
+# Case 2: Symbol is not declared in this region. In this case, we only support
+# non-CFGs at the moment (i.e. single block regions). We observe that in general
+# every symbol can be simplified within a block to have at most one fetch (start
+# of the block) at most one update (end of the block). This means that using
+# specification of an operation we can promote these symbols outside.
+# TODO: Add op promotion.
+
+
+def has_symbol(op: Operation) -> bool:
+    return isinstance(op, symref.Declare) or isinstance(
+        op, symref.Fetch) or isinstance(op, symref.Update)
+
+
+def get_symbol(op: Operation) -> str:
+    """
+    Returns a symbol attribute for an operation. If operation has no symbol
+    attributes, raises an exception.
+    """
+    if isinstance(op, symref.Declare):
+        return op.sym_name.data
+    elif isinstance(op, symref.Fetch) or isinstance(op, symref.Update):
+        return op.symbol.root_reference.data
+    else:
+        raise FrontendProgramException(
+            f"Operation '{op.name}' does not have symbol attributes.")
+
+
+def get_symbols(block: Block) -> Set[str]:
+    """Returns a set of all symbols defined/used in a basic block."""
+    symbols: Set[str] = set()
+    for op in block.ops:
+        if has_symbol(op):
+            symbols.add(get_symbol(op))
+    return symbols
+
+
+def lower_positional_bound(ops: List[Operation],
+                           op: Operation) -> Operation | None:
+    """
+    Returns an operation from `ops` which is a lower bound for `op`. If there is
```

Comment:
What does this means exactly?
Does this means finding the nearest operation before `op`? Could you change a bit the comments here?